Gravatar OAuth in SwiftUI

etoledom opened the pull request:
### Description
This PR implements the Gravatar OAuth flow.

![CleanShot 2024-08-22 at 14 51 23](https://github.com/user-attachments/assets/cfebb657-3b16-49f3-9ad4-5532e18bd426)


### Testing Steps

- Build and run the SwiftUI demo app
- Go to Profile Editor with oauth demo screen.
- insert an email and open the editor.
  - Check that OAuth flow works as expected.

etoledom committed:
Update .gitignore

etoledom committed:
Delete Secrets.swift

etoledom committed:
Update .gitignore

etoledom committed:
Adding Profile Editor view with oauth

etoledom commented on Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift:
```diff
@@ -0,0 +1,190 @@
+    }
+}
+
+enum OAuthError: Error {
+    case notConfigured
+    case couldNotCreateOAuthURLWithGivenSecrets
+    case couldNotGetCodeFromURL(String)
+    case oauthResponseError(String)
+    case unknown(Error)
+    case couldNotStoreToken(Error)
+}
```
This became a bit messy and I don't like it:

- `notConfigured`: Secrets are not set up. This maybe could be a `assertionFailure` instead of throwing? The idea is to warn the developer.
- `couldNotCreateOAuthURLWithGivenSecrets` this comes from using `Encodable` to add the query items to the URL. Probably if we add them manually we wouldn't need this error. And if this error happens, it's probably a developer mistake which should be catch on dev time (maybe assertion also?)
- `case couldNotGetCodeFromURL(String)` error on parsing the callback URL to get the code. This could be a runtime error but not sure how could this be handled 🤔 .
- `oauthResponseError(String)` Something went wrong at the backend. Needs to be handled in the UI.
- `unknown` ¯\_(ツ)_/¯  ... Probably on the UI also.
- `couldNotStoreToken(Error)` Keychain error. I got errors here because the keychan was wrongly set up. But it was a dev issue. Probably there are instances that this could be a user issue, but again, not sure how this should be handled. Technically we have the token and the user should be able to edit the profile, so I don't think that throwing is a good option. The problem is that they will need to be authenticated again next time.

pinarol replied:
I would add the assertionFailures for `notConfigured` and `couldNotCreateOAuthURLWithGivenSecrets` but not necessarily replace the "throw" statements. We still need the throws for prod. Throws also allow us return non-optional values.

pinarol replied:
> couldNotStoreToken(Error) Keychain error. I got errors here because the keychan was wrongly set up. But it was a dev issue. Probably there are instances that this could be a user issue, but again, not sure how this should be handled. Technically we have the token and the user should be able to edit the profile, so I don't think that throwing is a good option. The problem is that they will need to be authenticated again next time

As you said, since the user is able to continue I think we don't need to re-throw this one. This can be something we swallow inside a separate `do { ... } catch {...}` and handle just with an assertionFailure.

pinarol commented on Sources/GravatarUI/SwiftUI/OAuthSession/Keychain.swift:
```diff
@@ -0,0 +1,73 @@
+protocol SecureStorage: Sendable {
+    func setPassword(_ password: String, for key: String) throws
+    func deletePassword(with key: String) throws
+    func password(with key: String) throws -> String?
```
`password` is not quite the right terminology I think, how about `token` or `secret`? `password` is more for "something I know" kind of secrets. 🤔

etoledom replied:
I used password since is what we set in the keychain, though we use it to store an auth token.

On the implementation we use `kSecClass: kSecClassGenericPassword` (on line 22 for example).

Now, since it's for internal usage, I guess we can change to `secret` which I think is actually a more generic word than `password` ¯\_(ツ)_/¯

pinarol commented on Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift:
```diff
@@ -0,0 +1,190 @@
+        self.storage = storage
+    }
+
+    func isSession(with email: Email) -> Bool {
```
how about `hasSession(with: ...)`?

pinarol commented on Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift:
```diff
@@ -0,0 +1,190 @@
+enum OAuthError: Error {
+    case notConfigured
+    case couldNotCreateOAuthURLWithGivenSecrets
+    case couldNotGetCodeFromURL(String)
```
this was a bit ambiguous to me so I suggest:

> couldNotParseAccessCode

pinarol commented on Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift:
```diff
@@ -0,0 +1,190 @@
+        try? storage.password(with: email.rawValue)
+    }
+
+    func authenticate(with email: Email) async throws -> String {
```
It is not easily understandable what this method does and returns just by looking at the signature. `authenticate` is just half of OAuth. The work done here is I think described better by something like:

> acquireAccessToken(for email: Email) async throws -> String

etoledom replied:
Sounds good 👍

pinarol commented on Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift:
```diff
@@ -0,0 +1,190 @@
+        case let error as Keychain.KeychainError:
+            return .couldNotStoreToken(error)
+        default:
+            print("🛑 Error: \(type(of: error)) - \(error.localizedDescription)")
```
let's remove the "print" call. We can probably use the Apple’s os.log framework for logging, doesn't have to be in this PR though. 🤔

etoledom replied:
Sure, this was for debugging and see the types of the error thrown to be catch 👍 
`os.log` is a good idea. It might help developers debug issues on their users.

pinarol commented on Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift:
```diff
@@ -0,0 +1,190 @@
+    case couldNotGetCodeFromURL(String)
+    case oauthResponseError(String)
+    case unknown(Error)
+    case couldNotStoreToken(Error)
```
I think `DecodingError` deserves a separate case here.

pinarol commented on Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift:
```diff
@@ -0,0 +1,190 @@
+
+extension OAuthError {
+    static func from(error: Error) -> OAuthError {
+        switch error {
```
We can catch "DecodingError"s and re-throw them after adding an assertionFailure maybe? This means our decoding keys or types are not matching the response, so the developer should fix it.

```
case let error as DecodingError:
   assertionFailure("Unable to decode the response. Error: \(error.description)")
   throw .decodingError(error)

```

pinarol commented on Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift:
```diff
@@ -42,7 +42,7 @@ struct AvatarPickerView: View {
                     }
                 }
-                .task {
+                .onAppear() {
```
just curious, why is this change necessary?

etoledom replied:
Oh, I had a problem where the picker wasn't loading the first time. I thought this could be the problem. It wasn't and I forgot to change it back.

Also, `model.refresh()` is not async, and it opens a task itself, so I think `.task` here is not really necessary 🤔 

Though both options will work just fine.

pinarol replied:
Maybe we should make model.refresh() async. 🤔 

What's good about `.task` is, it will automatically cancel the task when the view is destroyed. `.onAppear()` does not have this feature. I think `.task` is slightly a better practice but we'll need to make model.refresh() async so the cancellation is propagated to child tasks I think 🤔. No need to dive into this in this PR though.

pinarol replied:
I created an issue to investigate this https://github.com/Automattic/Gravatar-SDK-iOS/issues/362

etoledom replied:
> What's good about .task is, it will automatically cancel the task when the view is destroyed.

Makes sense 🤔 

> I created an issue to investigate this https://github.com/Automattic/Gravatar-SDK-iOS/issues/362

Nice! Thanks 🙏

pinarol commented on Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift:
```diff
@@ -0,0 +1,52 @@
+                case .avatarPicker:
+                    AvatarPickerView(model: .init(email: email, authToken: token))
+                }
+                Button("Log out (for testing only)") {
```
This is meant to be a temp Button I guess but I feel like it can be useful during development in the long run as well. So how about we wrap it with some kind of compiler flag and keep it?

```
#if DEMO

#endif
```

etoledom replied:
Sounds good!

etoledom replied:
@pinarol - I tried this and the Build Conditions from the demo app won't be passed to the package, so there's no access to it from within the SDK.

I decided to move this button to the demo screen, previous to present the picker.

Now we will have a `Log out ` button when we detect a session:

![CleanShot 2024-08-22 at 14 28 18@2x](https://github.com/user-attachments/assets/c3f2fc24-4356-4d5f-8b86-a44338329c77)

pinarol replied:
I see. That works too. 👍

pinarol reviewed (commented):
Looks pretty good overall. Just dropping some non-critical comments and a question.

We need UI for this 2 case but it doesn't have to be in this PR.

- we don’t have the token yet but we’ll try (we need an activity indicating UI)
- we tried to get the token but we failed (we can reuse the ContentLoadingErrorView here as well)

etoledom committed:
Renaming SecureStorage password to secret

etoledom committed:
Showing loading indicator while asking for authorisation

etoledom committed:
Format

etoledom committed:
Fix tests build

etoledom committed:
Adding assertions

etoledom committed:
Merge branch 'trunk' into etoledom/swiftui-oauth

etoledom commented on Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift:
```diff
@@ -0,0 +1,58 @@
+        Task {
+            isAuthenticating = true
+            if !oauthSession.hasSession(with: email) {
+                _ = try? await oauthSession.retrieveAccessToken(with: email)
```
The possible errors are still not yet properly handled here, but we can do this alongside building the error screens to present these errors.

pinarol commented:
Something caught my attention, I see the diff for my secrets file:

<img width="388" alt="Screenshot 2024-08-22 at 15 59 07" src="https://github.com/user-attachments/assets/f20b03b9-17bc-44fd-a458-f62581f8b958">

Seems like the path doesn't match the paths in the `.gitignore`

etoledom commented:
@pinarol - I believe this is ready for another look 👀

etoledom commented:
> Something caught my attention, I see the diff for my secrets file:
> 
> <img alt="Screenshot 2024-08-22 at 15 59 07" width="388" src="https://private-user-images.githubusercontent.com/5032900/360438918-f20b03b9-17bc-44fd-a458-f62581f8b958.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQzMzE5NzgsIm5iZiI6MTcyNDMzMTY3OCwicGF0aCI6Ii81MDMyOTAwLzM2MDQzODkxOC1mMjBiMDNiOS0xN2JjLTQ0ZmQtYTQ1OC1mNjI1ODFmOGI5NTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjJUMTMwMTE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTk3MTlhYmE3OWMwNjIxMDNiYWMzNGZiNDljZjkxMDkwNDdmMGYzMTI4YWVjNTZlNjRkMDIzNWVhMmJmZTFmYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.eSk7XETIZyQh3xmnSqxy6qj9Qeo8y0lgZ3O3b2y5lmg">
> Seems like the path doesn't match the paths in the `.gitignore`

You can delete this file.

This was an error when we renamed the demo apps. We left this file which is not supposed to be in source control inside the old folder name. Now it shows because I removed its entry from .gitignore.

You get this since this file still exist in your machine, but it's safe to be deleted 👍

pinarol commented on Demo/Demo/Gravatar-SwiftUI-Demo/DemoProfileEditorViewView.swift:
```diff
@@ -0,0 +1,60 @@
+import SwiftUI
+import GravatarUI
```
typo in the filename ? "DemoProfileEditorViewView.swift"

pinarol commented on Sources/GravatarUI/SwiftUI/OAuthSession/Keychain.swift:
```diff
@@ -0,0 +1,73 @@
+struct Keychain: SecureStorage {
+    enum KeychainError: Error {
+        case unexpectedPasswordData
+        case cannotConvertPasswordIntoData
```
For consistency, let's replace `password` with `secret` here as well.

pinarol commented on Sources/GravatarUI/SwiftUI/View+Additions.swift:
```diff
@@ -25,3 +25,16 @@ extension View {
 }
+
+@MainActor
+extension View {
```
no need to add a new `extension View` I think? we already have one in the same file.

pinarol reviewed (approved):
Looks good, just leaving some nits. I don't need to check again.

pinarol commented:
> You get this since this file still exist in your machine, but it's safe to be deleted 👍

Oh I see. Cool.

etoledom committed:
Small review requests

etoledom committed:
Merge branch 'trunk' into etoledom/swiftui-oauth

etoledom committed:
Add `isPresented` property to ProfileEditor
